### PR TITLE
Add a way to do virtual merge through an API

### PR DESCRIPTION
**Pull request by jcoyne**
Ref https://github.com/sul-dlss/argo/issues/1458

### CODE REVIEW

File: app/services/constituent_service.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+# TODO: this should move to dor-services-app and dor-services-client
+# because we don't want Argo to have to know about the XML backing the model.
+#
+# Adds a constitutent relationship between a parent work and child works
```
Comment (mjgiarlo) [inline app/services/constituent_service.rb]: `s/constitutent/constituent/g`

File: app/controllers/objects_controller.rb [hunk 2]
```diff
@@ -28,6 +28,31 @@ def create
...
+    end
+
+    # Update the constituent relationship
+    begin
+      ConstituentService.new(parent_druid: params[:id]).add(child_druids: params[:constituent_ids])
+      head :no_content
+    rescue StandardError => e
+      return render json: {
+        errors: [
+          { title: 'bad request', detail: e }
```
Comment (mjgiarlo) [inline app/controllers/objects_controller.rb]: I wonder if this ought to be an internal server error 500 given this catches all sorts of exceptional behavior.

File: app/services/constituent_service.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+# TODO: this should move to dor-services-app and dor-services-client
```
Comment (mjgiarlo) [inline app/services/constituent_service.rb]: I think this TODO can be removed.

File: app/services/constituent_service.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+# because we don't want Argo to have to know about the XML backing the model.
+#
+# Adds a constituent relationship between a parent work and child works
+# by taking the followin actions:
+#  1. altering the contentMD of the parent
+#  2. add isConstituentOf assertions to the RELS-EXT of the children
+#  3. saving the parent and the child
```
Comment (mjgiarlo) [inline app/services/constituent_service.rb]: `s/child/child(ren)/` ?

File: app/controllers/objects_controller.rb [hunk 5]
```diff
@@ -28,6 +28,31 @@ def create
...
+      return render json: {
+        errors: [
+          { title: 'bad request', detail: 'Currently only constituent_ids are supported for update' }
+        ]
+      }, status: :bad_request
+    end
+
+    # Update the constituent relationship
+    begin
+      ConstituentService.new(parent_druid: params[:id]).add(child_druids: params[:constituent_ids])
```
Comment (mjgiarlo) [inline app/controllers/objects_controller.rb]: If only one value is in `constituent_ids`, will it still come across as an array?

File: app/services/constituent_service.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    end
+    child.add_relationship :is_constituent_of, parent
+    child.save!
+  end
+
+  def parent
+    @parent ||= WorkQueryService.find_modifiable_work(parent_druid)
+  end
+
+  attr_reader :parent_druid
```
Comment (mjgiarlo) [inline app/services/constituent_service.rb]: I have a mild preference for seeing this immediately below the `private` keyword but nbd.

File: app/services/reset_content_metadata_service.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+# Clears the contentMetadata datastream to the default, wiping out any members.
+class ResetContentMetadataService
+  def initialize(druid:, type: 'image')
+    @druid = druid
+    @type = type
+  end
+
+  def reset
```
Comment (mjgiarlo) [inline app/services/reset_content_metadata_service.rb]: Appreciate the rename from `purge` to `reset`!

File: app/controllers/objects_controller.rb (hunk 2 above)
Comment (ndushay) [reply app/controllers/objects_controller.rb]: I further wonder if the ConstituentService would throw a specific error, or set of them to be caught here?

File: app/services/constituent_service.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    end
+    parent.save!
+  end
+
+  private
+
+  def add_constituent(child_druid:)
+    child = WorkQueryService.find_modifiable_work(child_druid)
+    child.contentMetadata.ng_xml.search('//resource').each do |resource|
+      parent.contentMetadata.add_virtual_resource(child.id, resource)
```
Comment (ndushay) [inline app/services/constituent_service.rb]: I'm assuming there's good reason to leave the `add_virtual_resource` method in dor-services rather than port it here?  See dor-services/issues/652.   (Like -- we'd have to pull over all kinds of things to get the same functionality natively in this app?  or ... this app is just gating the access to dor-services gem?)

File: app/services/reset_content_metadata_service.rb (hunk 7 above)
Comment (ndushay) [reply app/services/reset_content_metadata_service.rb]: agree - nice.

File: app/services/constituent_service.rb (hunk 8 above)
Comment (ndushay) [inline app/services/constituent_service.rb]: I think we *might* want to keep the logging message for debugging?  https://github.com/sul-dlss/argo/blob/master/bin/virtual-merge#L33

I'm ok with removing it if @andrewjbtw is ok with it.  I understand it's on a diff VM than the one running the merge script ... just making sure we don't need the breadcrumb.

File: app/services/work_query_service.rb [hunk 9]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+# Responsible for retrieving information based on the given work (Dor::Item).
+class WorkQueryService
```
Comment (ndushay) [inline app/services/work_query_service.rb]: ooo - work is an overloaded term.   so is "object" ...  `DorItemQueryService` ?   `ItemQueryService`?  `ObjectQueryService` ?  

I'm concerned the existing naming will be confusing, esp. over time.

File: app/services/work_query_service.rb [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  # @param [#exists?, #find] work_relation - How we will query some of the related information
+  def initialize(id:, work_relation: default_work_relation)
+    @id = id
+    @work_relation = work_relation
+  end
+
+  delegate :allows_modification?, to: :work
+
+  # @raises [RuntimeError] if the item is not modifiable
+  def self.find_modifiable_work(druid)
```
Comment (ndushay) [inline app/services/work_query_service.rb]: is "find" the right concept here?   It's more like `assert_modifiable` or even `.modifiable?` or eve `.allows_modification?`, no?   

Same concern about naming --  this is really a method to assert that the item allows_modification;  I'm concern that the terms "find" and "work" will create confusion.

File: spec/requests/virtual_merge_spec.rb [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    allow(ConstituentService).to receive(:new).with(parent_druid: parent_id).and_return(service)
+  end
+
+  it 'merges the objects' do
+    put "/v1/objects/#{parent_id}",
+        params: { constituent_ids: [child1_id, child2_id] },
+        headers: { 'X-Auth' => "Bearer #{jwt}" }
+    expect(service).to have_received(:add).with(child_druids: [child1_id, child2_id])
+    expect(response).to be_successful
+  end
```
Comment (ndushay) [inline spec/requests/virtual_merge_spec.rb]: Add a test showing the ResetContentMetadataService.reset is called on the parent, as well?

File: spec/services/constituent_service_spec.rb [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+    context 'when the parent is open for modification' do
+      before do
+        add
+      end
+
+      it 'merges objects' do
+        expect(parent.contentMetadata.content).to be_equivalent_to <<~XML
+          <contentMetadata objectId="druid:parent1" type="image">
+            <resource sequence="1" id="#{namespaceless}_1" type="image">
```
Comment (ndushay) [inline spec/services/constituent_service_spec.rb]: I believe this is incorrect.  See:

https://github.com/sul-dlss-deprecated/dor-utils/issues/17#issuecomment-518770583 -- we are missing externalFile part.

or 

https://argo.stanford.edu/view/druid:bb258cw1178   contentMetadata

File: spec/services/constituent_service_spec.rb [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+              <relationship type="alsoAvailableAs" objectId="#{child1.id}"/>
+            </resource>
+            <resource sequence="2" id="#{namespaceless}_2" type="image">
+              <relationship type="alsoAvailableAs" objectId="#{child2.id}"/>
+            </resource>
+          </contentMetadata>
+        XML
+        expect(child1.object_relations[:is_constituent_of]).to eq [parent]
+        expect(child2.object_relations[:is_constituent_of]).to eq [parent]
```
Comment (ndushay) [inline spec/services/constituent_service_spec.rb]: it would be good to show this relationship doesn't exist before running the command.

File: app/services/constituent_service.rb [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+# by taking the followin actions:
+#  1. altering the contentMD of the parent
+#  2. add isConstituentOf assertions to the RELS-EXT of the children
+#  3. saving the parent and the child
+class ConstituentService
+  def initialize(parent_druid:)
+    @parent_druid = parent_druid
+  end
+
+  def add(child_druids:)
```
Comment (ndushay) [inline app/services/constituent_service.rb]: would this be clearer as "add_children"  or "add_constituents"?

File: spec/services/constituent_service_spec.rb [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+      it 'merges nothing' do
+        expect { add }.to raise_error RuntimeError
+        expect(parent.contentMetadata.content).to be_equivalent_to <<~XML
+          <contentMetadata />
+        XML
+        expect(child1.object_relations[:is_constituent_of]).to be_empty
+      end
+    end
+
+    context 'when the child is closed for modification' do
```
Comment (ndushay) [inline spec/services/constituent_service_spec.rb]: when _any_ child is close for modification?  Should test show second child closed for modification, and should it show what happens to parent contentMd in that case??

File: app/controllers/objects_controller.rb (hunk 2 above)
Comment (jcoyne) [reply app/controllers/objects_controller.rb]: I'm just going to drop the rescue block and let the Rails middleware + Honeybadger deal with any errors.  I don't see any benefit to specially handling the errors at this point.

File: app/services/work_query_service.rb (hunk 9 above)
Comment (jcoyne) [reply app/services/work_query_service.rb]: I don't like Item as it seems to be too generic (it is used in every domain).  "Work" is what Samvera has choosen and in the interest of aligning our practices with the greater community, that's what I'm inclined to choose.

File: spec/requests/virtual_merge_spec.rb (hunk 11 above)
Comment (jcoyne) [reply spec/requests/virtual_merge_spec.rb]: ResetContentMetadataService is called by ConstituentService, which we are checking for.

File: spec/services/constituent_service_spec.rb (hunk 12 above)
Comment (jcoyne) [reply spec/services/constituent_service_spec.rb]: The method is moving over the whole node, but the xml we're using as a fixture just doesn't have any children.

File: spec/services/constituent_service_spec.rb (hunk 13 above)
Comment (jcoyne) [reply spec/services/constituent_service_spec.rb]: We are showing that by creating the child fixture objects without any relationships at all.

File: app/services/constituent_service.rb (hunk 14 above)
Comment (jcoyne) [reply app/services/constituent_service.rb]: My thinking is that the ConstituentService class provides that context.

File: spec/services/constituent_service_spec.rb (hunk 15 above)
Comment (jcoyne) [reply spec/services/constituent_service_spec.rb]: I'm not sure I understand what you are asking for here.   Why does it matter which child is closed?

File: app/services/constituent_service.rb (hunk 8 above)
Comment (jcoyne) [reply app/services/constituent_service.rb]: Im trying to keep the changes to a manageable size. Moving add_virtual_resource is not required to gain the functionality added in this PR. 652 is still valid and can be done later.

File: app/services/work_query_service.rb (hunk 10 above)
Comment (jcoyne) [reply app/services/work_query_service.rb]: It is finding and ensuring that it’s modifable. I’m not sure I understand why this is a problem.

File: app/services/work_query_service.rb (hunk 9 above)
Comment (mjgiarlo) [reply app/services/work_query_service.rb]: Alignment with Samvera is generally fine by me and I'm accustomed to the "work" verbiage at this point. That said, the term we use most in other SDR codebases is "Object," in my reckoning, so I'd be inclined to switch to that if we're switching to anything. (I see a Dor::Item as a subclass of SDR "objects.")

Comment (jcoyne) [reply app/services/work_query_service.rb]: In the TACO work, we called it a DRO, so maybe we should go that direction? https://github.com/sul-dlss-labs/sdr3-models/blob/master/taco-maps/DRO.json

Comment (mjgiarlo) [reply app/services/work_query_service.rb]: @jcoyne I figured `Object` was defensible because DSA is Fedora-facing, and `Object` is standard Fedora nomenclature. I don't care *too* much; work, item, object, and DRO all work for me.

File: app/services/constituent_service.rb (hunk 8 above)
Comment (ndushay) [reply app/services/constituent_service.rb]: totally makes sense.  thanks for the explanation.

File: spec/services/constituent_service_spec.rb (hunk 15 above)
Comment (ndushay) [reply spec/services/constituent_service_spec.rb]: I was thinking it would be appropriate to show that the process ends even if a child in the middle is the one that can't be modified.  That is, showing a partially "merged" situation could be a more informative use case for future spelunking.

File: app/services/constituent_service.rb (hunk 14 above)
Comment (ndushay) [reply app/services/constituent_service.rb]: k.  certainly not something I'd go to the mat over.

File: spec/services/constituent_service_spec.rb (hunk 12 above)
Comment (ndushay) [reply spec/services/constituent_service_spec.rb]: It would be more useful to show the full "what we want it to look like when it is successful" here ... imagine trying to figure out if it's working right when you don't have a good example of what is supposed to happen in the tests.

File: app/controllers/application_controller.rb [hunk 16]
```diff
@@ -3,6 +3,14 @@
 class ApplicationController < ActionController::API
   include ActionController::MimeResponds
 
+  rescue_from ActionController::ParameterMissing do |exception|
+    render json: {
+      errors: [
+        { title: 'bad request', detail: exception.message }
+      ]
+    }, status: :bad_request
+  end
```
Comment (ndushay) [inline app/controllers/application_controller.rb]: I'm glad to see trapping for this problem and a better error, but ... should it be a separate commit, in a perfect world?  (no need to rework, just asking)

File: spec/services/constituent_service_spec.rb (hunk 12 above)
Comment (jcoyne) [reply spec/services/constituent_service_spec.rb]: That depends on the starting value of the publish attribute: https://github.com/sul-dlss/dor-services/blob/master/lib/dor/datastreams/content_metadata_ds.rb#L188  That varying behavior is tested in the dor-services gem.

File: spec/services/constituent_service_spec.rb (hunk 15 above)
Comment (jcoyne) [reply spec/services/constituent_service_spec.rb]: done
